Binary files ./RegMiner4APR-Regression-Bugs/WORKING/.gradle/6.6.1/executionHistory/executionHistory.bin and ./RegMiner4APR-Regression-Bugs/BIC/.gradle/6.6.1/executionHistory/executionHistory.bin differ
Binary files ./RegMiner4APR-Regression-Bugs/WORKING/.gradle/6.6.1/executionHistory/executionHistory.lock and ./RegMiner4APR-Regression-Bugs/BIC/.gradle/6.6.1/executionHistory/executionHistory.lock differ
Binary files ./RegMiner4APR-Regression-Bugs/WORKING/.gradle/6.6.1/fileHashes/fileHashes.bin and ./RegMiner4APR-Regression-Bugs/BIC/.gradle/6.6.1/fileHashes/fileHashes.bin differ
Binary files ./RegMiner4APR-Regression-Bugs/WORKING/.gradle/6.6.1/fileHashes/fileHashes.lock and ./RegMiner4APR-Regression-Bugs/BIC/.gradle/6.6.1/fileHashes/fileHashes.lock differ
Binary files ./RegMiner4APR-Regression-Bugs/WORKING/.gradle/6.6.1/fileHashes/resourceHashesCache.bin and ./RegMiner4APR-Regression-Bugs/BIC/.gradle/6.6.1/fileHashes/resourceHashesCache.bin differ
Binary files ./RegMiner4APR-Regression-Bugs/WORKING/.gradle/6.6.1/javaCompile/classAnalysis.bin and ./RegMiner4APR-Regression-Bugs/BIC/.gradle/6.6.1/javaCompile/classAnalysis.bin differ
Binary files ./RegMiner4APR-Regression-Bugs/WORKING/.gradle/6.6.1/javaCompile/jarAnalysis.bin and ./RegMiner4APR-Regression-Bugs/BIC/.gradle/6.6.1/javaCompile/jarAnalysis.bin differ
Binary files ./RegMiner4APR-Regression-Bugs/WORKING/.gradle/6.6.1/javaCompile/javaCompile.lock and ./RegMiner4APR-Regression-Bugs/BIC/.gradle/6.6.1/javaCompile/javaCompile.lock differ
Binary files ./RegMiner4APR-Regression-Bugs/WORKING/.gradle/6.6.1/javaCompile/taskHistory.bin and ./RegMiner4APR-Regression-Bugs/BIC/.gradle/6.6.1/javaCompile/taskHistory.bin differ
Binary files ./RegMiner4APR-Regression-Bugs/WORKING/.gradle/buildOutputCleanup/buildOutputCleanup.lock and ./RegMiner4APR-Regression-Bugs/BIC/.gradle/buildOutputCleanup/buildOutputCleanup.lock differ
Binary files ./RegMiner4APR-Regression-Bugs/WORKING/.gradle/buildOutputCleanup/outputFiles.bin and ./RegMiner4APR-Regression-Bugs/BIC/.gradle/buildOutputCleanup/outputFiles.bin differ
Binary files ./RegMiner4APR-Regression-Bugs/WORKING/.gradle/checksums/checksums.lock and ./RegMiner4APR-Regression-Bugs/BIC/.gradle/checksums/checksums.lock differ
Binary files ./RegMiner4APR-Regression-Bugs/WORKING/.gradle/checksums/sha1-checksums.bin and ./RegMiner4APR-Regression-Bugs/BIC/.gradle/checksums/sha1-checksums.bin differ
diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/adapter/MessagingMessageListenerAdapter.java ./RegMiner4APR-Regression-Bugs/BIC/spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/adapter/MessagingMessageListenerAdapter.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/adapter/MessagingMessageListenerAdapter.java	2025-12-24 04:11:05.095417071 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/adapter/MessagingMessageListenerAdapter.java	2025-12-24 04:07:12.639577750 +1100
@@ -34,6 +34,7 @@
 import org.springframework.lang.Nullable;
 import org.springframework.messaging.Message;
 import org.springframework.messaging.MessagingException;
+import org.springframework.messaging.handler.annotation.Header;
 import org.springframework.messaging.handler.annotation.Payload;
 import org.springframework.messaging.support.MessageBuilder;
 import org.springframework.remoting.support.RemoteInvocationResult;
@@ -333,12 +334,21 @@
 			for (int i = 0; i < this.method.getParameterCount(); i++) {
 				MethodParameter methodParameter = new MethodParameter(this.method, i);
 				/*
-				 * We're looking for a single non-annotated parameter, or one annotated with @Payload.
+				 * We're looking for a single parameter, or one annotated with @Payload.
 				 * We ignore parameters with type Message because they are not involved with conversion.
 				 */
+				boolean isHeader = methodParameter.hasParameterAnnotation(Header.class);
+				boolean isPayload = methodParameter.hasParameterAnnotation(Payload.class);
+				if (isHeader && isPayload) {
+					if (MessagingMessageListenerAdapter.this.logger.isWarnEnabled()) {
+						MessagingMessageListenerAdapter.this.logger.warn(this.method.getName()
+							+ ": Cannot annotate a parameter with both @Header and @Payload; "
+							+ "ignored for payload conversion");
+					}
+				}
 				if (isEligibleParameter(methodParameter)
-						&& (methodParameter.getParameterAnnotations().length == 0
-						|| methodParameter.hasParameterAnnotation(Payload.class))) {
+						&& (!isHeader || isPayload) && !(isHeader && isPayload)) {
+
 					if (genericParameterType == null) {
 						genericParameterType = extractGenericParameterTypFromMethodParameter(methodParameter);
 					}
diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/spring-rabbit/src/test/java/org/springframework/amqp/rabbit/annotation/EnableRabbitIntegrationTests.java ./RegMiner4APR-Regression-Bugs/BIC/spring-rabbit/src/test/java/org/springframework/amqp/rabbit/annotation/EnableRabbitIntegrationTests.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/spring-rabbit/src/test/java/org/springframework/amqp/rabbit/annotation/EnableRabbitIntegrationTests.java	2025-12-24 04:11:05.103417066 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/spring-rabbit/src/test/java/org/springframework/amqp/rabbit/annotation/EnableRabbitIntegrationTests.java	2025-12-24 04:07:12.647577744 +1100
@@ -131,6 +131,7 @@
 import org.springframework.transaction.annotation.EnableTransactionManagement;
 import org.springframework.transaction.annotation.Transactional;
 import org.springframework.util.ErrorHandler;
+import org.springframework.validation.annotation.Validated;
 
 import com.rabbitmq.client.Channel;
 import com.rabbitmq.http.client.Client;
@@ -1148,8 +1149,9 @@
 
 		}
 
-		@RabbitListener(id = "different", queues = "differentTypes", containerFactory = "jsonListenerContainerFactory")
-		public void handleDifferent(Foo2 foo) {
+		@RabbitListener(id = "different", queues = "differentTypes",
+				containerFactory = "jsonListenerContainerFactoryNoClassMapper")
+		public void handleDifferent(@Validated Foo2 foo) {
 			foos.add(foo);
 			latch.countDown();
 		}
@@ -1593,6 +1595,19 @@
 		}
 
 		@Bean
+		public SimpleRabbitListenerContainerFactory jsonListenerContainerFactoryNoClassMapper() {
+			SimpleRabbitListenerContainerFactory factory = new SimpleRabbitListenerContainerFactory();
+			factory.setConnectionFactory(rabbitConnectionFactory());
+			factory.setErrorHandler(errorHandler());
+			factory.setConsumerTagStrategy(consumerTagStrategy());
+			Jackson2JsonMessageConverter messageConverter = new Jackson2JsonMessageConverter();
+			factory.setMessageConverter(messageConverter);
+			factory.setReceiveTimeout(10L);
+			factory.setConcurrentConsumers(2);
+			return factory;
+		}
+
+		@Bean
 		public MeterRegistry meterRegistry() {
 			return new SimpleMeterRegistry();
 		}
@@ -2189,7 +2204,7 @@
 		}
 
 		@RabbitListener(queues = "test.converted.args2")
-		public String foo2a(@Payload Foo2 foo2, @Header("amqp_consumerQueue") String queue) {
+		public String foo2a(Foo2 foo2, @Header("amqp_consumerQueue") String queue) {
 			return foo2 + queue;
 		}
 
diff -r -U 3 ./RegMiner4APR-Regression-Bugs/WORKING/spring-rabbit/src/test/java/org/springframework/amqp/rabbit/listener/adapter/MessagingMessageListenerAdapterTests.java ./RegMiner4APR-Regression-Bugs/BIC/spring-rabbit/src/test/java/org/springframework/amqp/rabbit/listener/adapter/MessagingMessageListenerAdapterTests.java
--- ./RegMiner4APR-Regression-Bugs/WORKING/spring-rabbit/src/test/java/org/springframework/amqp/rabbit/listener/adapter/MessagingMessageListenerAdapterTests.java	2025-12-24 04:11:05.119417055 +1100
+++ ./RegMiner4APR-Regression-Bugs/BIC/spring-rabbit/src/test/java/org/springframework/amqp/rabbit/listener/adapter/MessagingMessageListenerAdapterTests.java	2025-12-24 04:07:12.659577736 +1100
@@ -52,7 +52,6 @@
 import java.util.Map;
 import org.springframework.amqp.utils.test.TestUtils;
 import org.springframework.messaging.handler.annotation.Headers;
-
 /**
  * @author Stephane Nicoll
  * @author Artem Bilan
@@ -388,11 +387,10 @@
 		public String failWithReturn(Integer input) {
 			throw new IllegalArgumentException("Expected test exception");
 		}
-		
+
 		@SuppressWarnings("unused")
 		public void withHeaders(Foo foo, @Headers Map<String, Object> headers) {
 		}
-
 	}
 
 	private static class Foo {
